File: frontend/src/types/factor.ts
export interface FactorCatalog {
  code: string;
  name: string;
  name_cn?: string;
  category: string;
  sub_category?: string;
  description?: string;
  formula?: string;
  source?: string;
  polarity: string;
  data_requirements?: string[];
  parameters?: Record<string, unknown>;
  paper_reference?: string;
  status: string;
  popularity: number;
}

export interface FactorDetail extends FactorCatalog {
  recent_values?: FactorValue[];
  ic_summary?: Record<string, number>;
}

export interface FactorValue {
  ts_code: string;
  trade_date: string;
  raw_value: number;
  normalized_value: number;
  percentile?: number;
}

export interface FactorCategory {
  category: string;
  count: number;
}

File: backend/app/schemas/factor.py
from pydantic import BaseModel
from typing import Optional, Any
from datetime import date


class FactorCatalogOut(BaseModel):
    code: str
    name: str
    name_cn: Optional[str] = None
    category: str
    sub_category: Optional[str] = None
    description: Optional[str] = None
    formula: Optional[str] = None
    source: Optional[str] = None
    polarity: str
    data_requirements: Optional[Any] = None
    parameters: Optional[Any] = None
    paper_reference: Optional[str] = None
    status: str = "active"
    popularity: int = 0

    class Config:
        from_attributes = True


class FactorCategoryOut(BaseModel):
    category: str
    count: int


class FactorValuesOut(BaseModel):
    factor_code: str
    ts_code: str
    trade_date: date
    raw_value: Optional[float] = None
    normalized_value: Optional[float] = None
    percentile: Optional[float] = None

    class Config:
        from_attributes = True


class FactorComputeRequest(BaseModel):
    factor_code: str
    start_date: date
    end_date: date


class FactorDetailOut(FactorCatalogOut):
    recent_values: Optional[list] = None
    ic_summary: Optional[dict] = None
